File: src/tables/hmtx.cpp
#include "tables.h"

void parseHmtx(const quint16 numberOfMetrics, const quint16 numberOfGlyphs, Parser &parser)
{
    parser.readArray("Metrics", numberOfMetrics, [&](const auto index){
        parser.beginGroup(index);
        parser.read<UInt16>("Advance width");
        parser.read<Int16>("Left side bearing");
        parser.endGroup();
    });

    if (numberOfGlyphs <= numberOfMetrics) {
        return;
    }

    parser.readArray("Additional Metrics", numberOfGlyphs - numberOfMetrics, [&](const auto index){
        parser.beginGroup(numberOfMetrics + index);
        parser.read<Int16>("Left side bearing");
        parser.endGroup();
    });
}

File: core/src/tables/hmtx.rs
use crate::parser::*;
use crate::Result;

pub fn parse(number_of_metrics: u16, number_of_glyphs: u16, parser: &mut Parser) -> Result<()> {
    for i in 0..number_of_metrics {
        parser.begin_group(format!("Glyph {}", i));
        parser.read::<u16>("Advance width")?;
        parser.read::<i16>("Left side bearing")?;
        parser.end_group();
    }

    for i in 0..(number_of_glyphs - number_of_metrics) {
        parser.begin_group(format!("Glyph {}", number_of_metrics + i));
        parser.read::<i16>("Left side bearing")?;
        parser.end_group();
    }

    Ok(())
}
